core/chains/evm/forwarders/forwarder_manager.go

```diff
@@ -0,0 +1,308 @@
+		return nil, errors.Wrap(err, "Failed to pack forwarder payload")
+	}
+
+	dataBytes := append(forwardABI.ID, callArgs...)
+	return dataBytes, nil
+}
+
+func (f *FwdMgr) getContractSenders(addr common.Address) ([]common.Address, error) {
+	if senders, ok := f.getCachedSenders(addr); ok {
+		return senders, nil
+	}
+	senders, err := f.getAuthorizedSenders(addr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to call getAuthorizedSenders on %s", addr)
+	}
+	f.setCachedSenders(addr, senders)
+	f.subscribeSendersChangedLogs(addr)
+	return senders, nil
+}
+
+func (f *FwdMgr) getAuthorizedSenders(addr common.Address) ([]common.Address, error) {
+	c, err := authorized_receiver.NewAuthorizedReceiverCaller(addr, f.evmClient)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to init forwarder caller")
+	}
+	opts := bind.CallOpts{Context: f.ctx, Pending: false}
+	senders, err := c.GetAuthorizedSenders(&opts)
+	if err != nil {
+		return nil, err
+	}
+	return senders, nil
+}
+
+func (f *FwdMgr) initForwardersCache(ctx context.Context, fwdrs []Forwarder) {
+	for _, fwdr := range fwdrs {
+		senders, err := f.getAuthorizedSenders(fwdr.Address)
+		if err != nil {
+			f.logger.Warnw("Failed to call getAuthorizedSenders on forwarder", fwdr, "err", err)
+			continue
+		}
+		f.setCachedSenders(fwdr.Address, senders)
+
+	}
+}
+func (f *FwdMgr) subscribeForwardersLogs(fwdrs []Forwarder) {
+	for _, fwdr := range fwdrs {
+		f.subscribeSendersChangedLogs(fwdr.Address)
+	}
+}
+
+func (f *FwdMgr) subscribeSendersChangedLogs(addr common.Address) {
+	f.logpoller.MergeFilter(
+		[]common.Hash{authorized_receiver.AuthorizedReceiverAuthorizedSendersChanged{}.Topic()},
+		addr)
+}
+
+func (f *FwdMgr) setCachedSenders(addr common.Address, senders []common.Address) {
+	f.cacheMu.Lock()
+	defer f.cacheMu.Unlock()
+	f.sendersCache[addr] = senders
+}
+
+func (f *FwdMgr) getCachedSenders(addr common.Address) ([]common.Address, bool) {
+	f.cacheMu.RLock()
+	defer f.cacheMu.RUnlock()
+	addrs, ok := f.sendersCache[addr]
+	return addrs, ok
+}
+
+func (f *FwdMgr) runLoop() {
+	defer f.wg.Done()
+	tick := time.After(0)
+
+	for {
+		select {
+		case <-tick:
+			tick = time.After(utils.WithJitter(authLogCheckInterval))
+			addrs := f.collectAddresses()
+			logs, err := f.logpoller.LatestLogEventSigsAddrsWithConfs(f.latestBlock, authTopics, addrs, 0)
+			if err != nil {
+				f.logger.Warnw("Failed to retrieve latest log round", "err", err)
+				continue
+			}
+			if len(logs) == 0 {
+				f.logger.Debugf("Empty auth update round for addrs: %s, skipping", addrs)
+				continue
+			}
+
+			f.logger.Debugf("Handling new %d auth updates", len(logs))
+
+			for _, log := range logs {
+				if err = f.handleAuthChange(log); err != nil {
+					f.logger.Warnf("Error handling auth change %s: %s", log.TxHash, err)
+				}
+			}
+
+		case <-f.chStop:
+			return
+		}
+	}
+}
+
+func (f *FwdMgr) handleAuthChange(log evmlogpoller.Log) error {
+	if f.latestBlock > log.BlockNumber {
```

Comment:
in steady state, we'll continue to get the same logs right so doesn't this need to be >=

Reply:
Once we consume logs, we don't want to process them again (they're supposed to be in mem)

Reply:
My bad, didn't notice that. Done